libbeat/outputs/shipper/shipper.go

```diff
@@ -124,19 +129,36 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
 		meta, err := convertMapStr(e.Content.Meta)
 		if err != nil {
 			// conversion errors are not recoverable, so we have to drop the event completely
-			c.log.Errorf("%d/%d failed to convert event metadata to protobuf: %w", i+1, len(events), err)
+			c.log.Errorf("%d/%d failed to convert event metadata to protobuf, dropped: %w", i+1, len(events), err)
 			dropped++
 			continue
 		}
 
 		fields, err := convertMapStr(e.Content.Fields)
 		if err != nil {
-			c.log.Errorf("%d/%d failed to convert event fields to protobuf: %w", i+1, len(events), err)
+			c.log.Errorf("%d/%d failed to convert event fields to protobuf, dropped: %w", i+1, len(events), err)
+			dropped++
+			continue
+		}
+
+		uuid, err := uuid.DefaultGenerator.NewV4()
+		if err != nil {
+			// this is unrecoverable and happens only when the random generator fails
+			batch.Cancelled()
+			return fmt.Errorf("failed to generate an event UUID: %w", err)
+		}
+		eventID := uuid.String()
```

Comment:
You shouldn't need to generate a unique ID here, the queue ID in the reply will give you one back. The event ID here is mostly intended for the case where an input has some natural unique ID that is more convenient to use than the queue ID or generating a random UUID.

My primary concern with keeping this here all the time would be the performance penalty of invoking what is probably the cryptographic random number generator for every event, and UUID comparisons being string comparisons rather than plain integers. There are no benchmarks for this yet though.

This is probably convenient for testing though, I'm just not sure this should be the default ID strategy for every event sent from any beat.

Reply:
@cmacknz how am I suppose to know which [`EventResult`](https://github.com/elastic/elastic-agent-shipper/blob/f2aa770912437213b866ee978692c0f0dc7af572/api/shipper.proto#L76-L87) is for which event that I sent?

I don't see any other way how to identify them. Is it supposed to always keep the same list order? What about partial acceptance (when the queue is full), etc.

Reply:
I believe I wrote the specification that the order of response is supposed to match the order in the request. We haven't tried to implement that yet though, so I am not sure if that is a reasonable approach yet.

You can assign an event ID for testing, but I don't think it should be automatically added by the shipper output here.

Reply:
I removed the retries based on IDs in https://github.com/elastic/beats/pull/31558/commits/40e4d3fb5f9c0117f1482148639464b927f8539b and now it's based purely on the ordering. So, we need to keep it in mind when implementing the server.